# Primer reto - Uso de variables

## 1. Contexto

Juan vive en una finca y tiene un arbol de frutas, el sueño de Juan es un amigo que le ayude a recoger las frutas cuando caen y se los traiga a sus manos. Juan construyó un robot y le puso de nombre Kevi.

Juan usa YarpTp para controlar a Kevi, pero tiene una duda, ¿Cómo se debe programar a Kevi para que busque la fruta y regrese al punto de origen? Mira la siguiente gráfica para tener más información.

## 2. Resultado esperado

Lograr que Kevi avance de un punto A a un punto B en línea recta.


## 3. Resultados de aprendizaje

* Aprender sobre uso de variables y expresiones.


## 4. Historias de usuario

* HU01: Moverse hacia adelante
    - Como Kevi
    - Quiero moverme hacia adelante una cantidad x de segundos
    - Para alcanzar un objetivo
* HUS02: Darse vuelta y regresar x segundos
    - Como Kevi
    - Quiero girar 180 grados y moverme de nuevo la cantidad x de segundos
    - Para regresar a mi punto de origen
 
 
##  5. Datos, información, variables y estructuras de datos

Para dar solución a este caso, debemos analizar los datos a utilizar, es conveniente seleccionar los datos en (datos de entrada y datos de salida):

### Datos de entrada: 

* Tiempo de desplazamiento.
* Velocidad de desplazamiento.
* dirección de desplazamiento.

Se consideran datos de entrada porque es información necesaria para realizar los procedimientos que permiten entregar la solución esperada o el resultado esperado. Para poder manipular estos datos en un Algoritmo o en un Programa de computador, es necesario definir las variables en donde se almacenarán los datos.

### Procedimiento para definir variables:
1. Tipo de Datos (numéricos o alfanumérico)
2. Defina los nombres de cada una de las variables (tenga en cuenta las normas para nombrar variables)
    #### Normas para nombres de variables
    * No espacios en blanco
    * No caracteres extraños (* + ! = ? ¡ ? ~ ^ { })
    * Use nombres que se relacionen con el dato que almacena
    * Puede utilizar letras y números

## 6. Estrategia de solución:
A continuación damos solución al problema teniendo en cuenta las etapas necesarias para el Ciclo de vida de Software

### A. Análisis:
En esta etapa se debe comprender de forma detallada cual es la problemática a resolver, comprobando el dominio o contexto en el cual se encuentra dicho problema. Al realizar la síntesis de las estructuras de datos y los requerimiento funcionales se está obteniendo la información necesaria y suficiente para afrontar su respectiva solución.

### B. Diseño:
Después de obtener la información necesaria para solucionar la problemática a resolver, se debe especificar las estrategias para dar una solución eficiente.

A continuación se encuentran el conjunto de actividades para dar solución al problema planteado.


In [ ]:
# El dato digitado es almacenado en la variable tiempo.

tiempo = input("Ingrese el tiempo de desplazamiento:")


In [ ]:
# Se almacena la velocidad de desplazamiento en la variable velocidad.

velocidad = input("Ingrese la velocidad de desplazamiento:")


Después de capturar los datos de entrada y almacenados en sus respectivas variables se procederá a las operaciones teniendo en cuenta el enunciado:

>  En determinado tiempo el robot gira 180° y vuelve a avanzar.

Para la realización de esta tarea haremos uso de los componentes de la libreria de YarpTp.

In [ ]:
from YarpTp import YarpTp
Y = YarpTp()

Para el movimiento hacia adelante tenemos **Forward(tiempo, velocidad)** que recibe como parametro los datos ingresados con anterioridad.


In [ ]:
Y.Forward(tiempo,velocidad)

para giros de 90° hacia la derecha **TurnRight(tiempo)**.

In [ ]:
Y.TurnRight(tiempo)

para giros de 90° hacia la izquierda tenemos **TurnLeft(tiempo)**

In [ ]:
Y.TurnLeft(tiempo)

### C. Implementación:

De esa manera podemos resover la historia de usuario **HUS01**:

In [ ]:
Y = YarpTp()
tiempo = input("Ingrese el tiempo de desplazamiento:")
velocidad = input("Ingrese la velocidad de desplazamiento:")
Y.Forward(tiempo,velocidad)

Y de esta manera resolvemos la historia de usuairo **HUS02**:

In [ ]:
tiempo_giro = input("Ingrese el tiempo de giro:")
Y.TurnRight(tiempo_giro)
Y.Forward(tiempo,velocidad)

Entonces, la implementación completa de este reto seria:

In [12]:
from YarpTp import YarpTp

Y = YarpTp()
tiempo = input("Ingrese el tiempo de desplazamiento:")
velocidad = input("Ingrese la velocidad de desplazamiento:")
tiempo_giro = input("Ingrese el tiempo de giro:")
Y.Forward(tiempo,velocidad)
Y.TurnRight(tiempo_giro)
Y.Forward(tiempo,velocidad)


Ingrese el tiempo de desplazamiento:1
Ingrese la velocidad de desplazamiento:80
Ingrese el tiempo de giro:3.2



<center><video src="leccion.mp4" width="420" height="400" Controls muted></video></center>

Una función puede ser llamada desde cualquier parte del código si está previamente definida.

### D. Pruebas:
Después de codificar las actividades y/o procesos en la la herramienta computacional seleccionada, es necesario realizar la debidas pruebas que garanticen el correcto funcionamiento del programa construido bajo un conjunto de situaciones posibles a las que se pueda enfrentar.

<table>
<tr>
<th>Datos de Velocidad</th>
<th>Tiempos de desplazamiento(sg)</th>
<th>Tiempos de giro(sg)</th>
</tr>
<tr>
<td>40</td>
<td>2.0</td>
<td>1.0</td>
</tr>
<tr>
<td>70</td>
<td>1.0</td>
<td>0.8</td>
</tr>
<tr>
<td>100</td>
<td>0.5</td>
<td>0.5</td>
</tr>
<tr>
<td>80</td>
<td>2.0</td>
<td>3.2</td>
</tr>
</table>
